test/test_dynamo_cudagraphs.py

```diff
@@ -0,0 +1,194 @@
+# Owner(s): ["module: cuda graphs"]
+
+import sys
+import torch
+from torch.testing._internal.common_utils import (
+    TestCase,
+    run_tests,
+)
+
+from unittest.mock import patch
+import functools
+
+try:
+    import torchdynamo
+    from torch.cuda._dynamo_graphs import aot_autograd_cudagraphs
+    import functorch  # noqa: F401
+
+    TEST_DYNAMO = True
+except ImportError:
+    TEST_DYNAMO = False
+
+TEST_CUDA = torch.cuda.is_available()
+
+if not TEST_CUDA or not TEST_DYNAMO:
+    print("CUDA or dynamo not available, skipping tests", file=sys.stderr)
+    TestCase = object  # noqa: F811
+
+
+def composed(*decs):
+    def deco(f):
+        for dec in reversed(decs):
+            f = dec(f)
+        return f
+
+    return deco
+
+
+def assert_aot_autograd_counter(ok=True):
+    def deco(f):
+        @functools.wraps(f)
+        def wrap(self, *args, **kwargs):
+            torchdynamo.utils.counters.clear()
+            r = f(self, *args, **kwargs)
+            c_ok = torchdynamo.utils.counters["aot_autograd"]["ok"]
+            c_not_ok = torchdynamo.utils.counters["aot_autograd"]["not_ok"]
+            if ok:
+                self.assertGreater(c_ok, 0)
+                self.assertEqual(c_not_ok, 0)
+            else:
+                self.assertEqual(c_ok, 0)
+                self.assertGreater(c_not_ok, 0)
+            return r
+
+        return wrap
+
+    return deco
+
+
+def patch_all(ok=True):
+    return composed(
+        patch("torchdynamo.config.verify_correctness", True),
+        assert_aot_autograd_counter(ok),
+    )
+
+
+N_ITERS = 5
+
+
+class TestDynamoCudaGraphs(TestCase):
+    @patch_all()
+    def test_basic(self):
+        def model(x, y):
+            return (x + y) * y
+
+        with torchdynamo.optimize(aot_autograd_cudagraphs):
```

Comment:
n00b q: Is the optimization unrolled outside of the `with` scope? If you called `torchdynamo.optimize()` in a loop would the result be the same as calling it once?

Reply:
this is a little subtle

First, the context manager is misleading. It doesn't actually turn on optimization for the inside of the manager. Optimization only turns on when you hit a new frame (e.g., do a function call).

With that out of the way, what if you have the optimization inside or outside of a loop? It will depend. If the loop successfully unrolls, then you will get a compiled graph outside the loop that has the unrolled graph. But let's say there's some reason we can't compile the outer frame. Then we will compile the inner function, and the two applications are equivalent/